### PR TITLE
BUG: mangle_dup_cols in read_csv replacing existing cols when conflict with target col

### LINKED ISSUE #14704: read_csv(...,mangle_dupe_cols=True) causes silent data loss for certain column names. Request introduction of mangle_dupe_cols_str
Lets say we have data with both column and row headers. Lets say this dataset also simply outputs a ```0``` in that csv file for the cell at the intersection of the column and row headers (A1 in excel notation, or cell (0,0) in the csv file). Additionally, both ```"0"``` and ```"0.1"``` are valid column names:

RH\\CH|-0.1|0|0.1
---|---|---|---
10.0|123|456|789
20.0|012|345|678

Thus, "RH\\CH" is replaced by ```"0"``` on export (of which I have no control). 

```python
import pandas as pd
from cStringIO import StringIO

data = '0,-0.1,0,0.1\n10.0,123,456,789\n20.0,012,345,678'
df = pd.read_csv(StringIO(data),index_col=0,header=0)
print(df)

      -0.1  0.1  0.1
0                   
10.0   123  789  789
20.0    12  678  678

```

The name mangling will change the duplicate ```"0"``` column containing data to ```"0.1"```, thus if a real data column has the name ```"0.1"```, this data will be copied (?) back to the mangled duplicate ```"0"``` column. Additionally, if the true ```"0.1"``` data name is missing, then this name mangling is frustrating as there is no obvious way to determine if the now present ```"0.1"``` column is a duplicate ```"0"``` that has been mangled or is a real ```"0.1"``` data series.

I propose the addition of a ```mangle_dupe_cols_str``` keyword option that defaults to ```'.'``` to preserve the current behavior. However, it can be passed as a kwarg to ```read_csv``` in cases where the period name mangling could result in further duplicate columns.

In https://github.com/pandas-dev/pandas/blob/v0.19.1/pandas/io/parsers.py#L2109-L2111:
```python
                        if cur_count > 0:
                            this_columns[i] = '%s.%d' % (col, cur_count)
                        counts[col] = cur_count + 1
```
should be adapted to 
```python
                        if cur_count > 0:
                            this_columns[i] = '%s%s%d' % (col, self.mangle_dupe_cols_str, cur_count)
                        counts[col] = cur_count + 1
```
and lines 1042-1043 in https://github.com/pandas-dev/pandas/blob/v0.19.1/pandas/io/parsers.py#L1042-L1043 changed to:
```python
        self.mangle_dupe_cols = kwds.get('mangle_dupe_cols', True)
        self.mangle_dupe_cols_str = kwds.get('mangle_dupe_cols_str', '.')        
        self.infer_datetime_format = kwds.pop('infer_datetime_format', False)
```

#### Expected Output
Then the corrected output would be:
```python
>>> data = '0,-0.1,0,0.1\n10.0,123,456,789\n20.0,012,345,678'
>>> df = pd.read_csv(StringIO(data),index_col=0,header=0,mangle_dupe_cols_str='__')
>>> print(df)

      -0.1  0__1  0.1
0                   
10.0   123  456  789
20.0    12  345  678

```
where subsequent operations could identify any columns with the mangle_dupe_cols_str, if needed. The current behavior silently causes data loss for certain column names.

<details>
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en_US

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: None
Cython: 0.24.1
numpy: 1.11.1
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.4.6
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: 1.1.0
tables: 3.2.2
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.3
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.8
boto: 2.42.0
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #14704
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

This ensures that already existing column names are kept. The index of the duplicated col is one higher in this case

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -657,46 +657,56 @@ cdef class TextReader:
...
+                    this_header.append(name)
 
-                    count = counts.get(name, 0)
+                if not self.has_mi_columns and self.mangle_dupe_cols:
+                    col_loop_order = [i for i in range(len(this_header))
+                                      if i not in unnamed_col_indices
+                                      ] + unnamed_col_indices
+                    counts = {}
+
+                    for i in col_loop_order:
```
Comment (jreback) [inline pandas/_libs/parsers.pyx:686]: can you add some comments about what this is doing

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@gfyoung if any comments

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (phofl) [reply pandas/_libs/parsers.pyx:686]: Added

**Review by gfyoung — APPROVED** [review]
Nice work!

The modified expected results for the tests will take a little getting used to, but they of course make total sense.

**Comment by phofl** [comment]
Thx

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
I think its worth making this a sub-section in the whatsnew docs to highlite it.

**Comment by phofl** [comment]
Added the sub section

**Comment by phofl** [comment]
@jreback greenish

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Comment by jreback** [comment]
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=69777&view=logs&j=cb04ab33-336b-54ee-68fe-dcd3abde835b&t=0558d00d-ff4a-5538-0de9-a0e7f449700f

hmm this actually be related

**Comment by phofl** [comment]
Are you sure? previous commit was green, only merged master and modified whatsnew afterwards.

**Comment by jreback** [comment]
not sure at all - if it's not a problem on master then ignore

**Comment by phofl** [comment]
Have seen similar things in the past like https://github.com/pandas-dev/pandas/runs/4331169003, hence assumed this would be unrelated. Lets see if it passes on master

**Pull request merged**